Julia.sublime-syntax

```diff
@@ -25,585 +25,1088 @@ variables:
+        - include: escaped-char
+    - match: "```"
+      scope: punctuation.definition.string.begin.julia
+      push:
+        - meta_scope: meta.string.julia string.interpolated.julia
+        - match: "```"
+          scope: punctuation.definition.string.end.julia
           pop: true
+        - include: interpolated-julia
+        - include: escaped-char
 
-  literals:
+  constants:
     - match: |-
         (?x)
+        (\b|(?=.))
         (?: # Dashes betwen numeric symbols (11 = 1_1) are allowed everywhere.
-          {{b}}0b[0-1](?:_?[0-1])*|             # binary
-          {{b}}0o[0-7](?:_?[0-7])*|             # octal
-          {{b}}0x[\da-fA-F](?:_?[\da-fA-F])*|   # hex
-          {{b}}(?:
+          0b[0-1](?:_?[0-1])*|             # binary
+          0o[0-7](?:_?[0-7])*|             # octal
+          0x[\da-fA-F](?:_?[\da-fA-F])*|   # hex
+          (?:
             \.\d(?:_?\d)*|                      # .11, .11
             \d(?:_?\d)*(?:\.(?:\d(?:_?\d)*)?)?  # 11.11, 11., 11
           )
           (?:e[-+]?\d(?:_?\d)*)?                # Any of the above followed by e+123 or similar, for scientific notation.
         )
       scope: constant.numeric.julia
+      push: after-expression
     - match: \b(true|false|nothing|missing|ℯ|pi|π|im|undef|NaN|NaN16|NaN32|NaN64|Inf|Inf16|Inf32|Inf64|ARGS|C_NULL|ENDIAN_BOM|ENV|LOAD_PATH|PROGRAM_FILE|STDERR|STDIN|STDOUT|VERSION)\b
       scope: constant.language.julia
+      push: after-expression
 
-  type-annotation:
-    # Dollar is ok because types can be interpolated.
-    # Dot is ok because types can be picked from modules,
-    # but no more than one dot, because splat can follow type.
-    - match: (::)\s*((?:(?!\.\.)(?:[$.]|{{symb_id}}))*({{nested_curly_and_round_sloppy}})?)\s*(where)\s+
+  symbols:
+    - match: '(:)(?!:)({{symb_id}}|\.?{{long_op}}|\.?{{symb_op}}|\.)'
       captures:
-        1: keyword.operator.julia
-        2: support.type.julia
-        4: keyword.other.julia
-      push: where-type
-    - match: (::|<:|>:)\s*((?:(?!\.\.)(?:[$.]|{{symb_id}}))*({{nested_curly_and_round_sloppy}})?)
+        1: keyword.operator.colon.julia constant.other.symbol.julia
+        2: constant.other.symbol.julia
+    - match: '(:)(\()'
       captures:
-        1: keyword.operator.julia
-        2: support.type.julia
+        1: keyword.operator.colon.julia constant.other.expression.julia
+        2: punctuation.section.expression.begin.julia
+      push:
+        - meta_scope: meta.expression.julia
+        - match: \)
+          scope: punctuation.section.expression.end.julia
+          set: after-expression
+        - include: interpolated-julia
+        - include: main
 
-  type-comparison:
-    - match: '{{type_comparison_regex}}'
-      captures:
-        1: support.type.julia
-        3: keyword.operator.julia
-        4: support.type.julia
+  macros:
+    - match: '@{{symb_id}}'
+      scope: support.function.julia
+    - match: '@\.'
+      scope: support.function.julia
```

Comment:
Was it intended to remove the additional `variable.macro` scope? I think it is nice to give users the option to highlight macros differently from function calls.
Same in [L121](https://github.com/JuliaEditorSupport/Julia-sublime/blob/73a881fdc5876f599c28598366de1fcf8f235dc3/Julia.sublime-syntax#L121).

Reply:
It was not intended. I agree that it’s a good idea to have those scopes back.

Reply:
It was not intended. I agree that it’s a good idea to have those scopes back.

Reply:
With the changes from your last commit (`variable.macro.julia support.function.julia`) it would only work if the color scheme targets this full scope, but a rule for only `variable.macro` will be overriden by a `support` or `support.function` rule in the scheme. Therefore better change the order to `support.function.julia variable.macro.julia` like it was before, so that it will still work if someone used only `variable.macro`?

Reply:
I was trying to match the scoping convention of python's `@property` decorator.

<img width="415" alt="Screen Shot 2020-01-09 at 11 54 50 AM" src="https://user-images.githubusercontent.com/1690993/72100137-f13e3f00-32d6-11ea-8579-71bea9b1ea9f.png">

Another possibility is to remove `support.function.julia` because technically the most macros are not built-in.

Reply:
I would say Julia macros are more like preprocessor directives in C/C++/C#, than Python's `@...` decorators or for example `@override` annotations in Java. To remove `support.function.julia` would be an option, but if the color scheme doesn't target `variable` or `variable.macro` they wouldn't be colored at all. In fact that would be the case for e.g. the Monokai and Mariana scheme. So I think to use `support.function.julia variable.macro.julia` would be the better option. 

Another alternative could be to use `support.macro.julia` for all built-in macros ([list](https://github.com/JuliaLang/julia/blob/45a0381baf5555952caea0ccb43b377ac5ec8ae2/base/exports.jl#L925)) and `variable.macro.julia` for the rest. On the other hand this would mean that common macros from imported packages would only get `variable.macro` then as well.

Reply:
After reading a bit more about Python decorators, I recognized that they are more powerful than I initially thought. Howerver, for a non-predefined decorator like `@my_decorator`, the Python syntax uses only `variable.annotation.python meta.generic-name.python`. So I would recommend to either remove the `support` scope for Julia macros that are not built-in, or change the scope order to how it was before.

Reply:
I have removed the scope `support.function` for non builtin macros.

Reply:
Alright, I think that is a good solution. The only thing that seems to be missed now is `@.`.
```diff
-    - match: '(@)({{base_macros}})'
+    - match: '(@)({{base_macros}}|\.)'
```

Reply:
I thought @. has been deprecated, isn’t it?

Reply:
Looks like I was wrong.